Fix AppSec crash when parsing integer http headers

vpellan opened the pull request:
**What does this PR do?**
<!-- A brief description of the change being made with this pull request. -->
This fixes a crash when malformed headers are sent (e.g. headers with numbers)

**Motivation:**
<!-- What inspired you to submit this pull request? -->
This fixes issue #3782 

**Additional Notes:**
<!-- Anything else we should know when reviewing? -->

**How to test the change?**
<!--
Describe here how the change can be validated.
You are strongly encouraged to provide automated tests for this PR (unit or integration).
If this change cannot be feasibly tested, please explain why,
unless the change does not modify code (e.g. only modifies docs, comments).
-->

`bundle exec appraisal ruby-x.x-rack-x rake spec:appsec:rack`

p-datadog commented on lib/datadog/appsec/contrib/rack/gateway/request.rb:
```diff
@@ -41,7 +41,7 @@ def method
             def headers
               result = request.env.each_with_object({}) do |(k, v), h|
-                h[k.gsub(/^HTTP_/, '').downcase!.tr('_', '-')] = v if k =~ /^HTTP_/
+                h[k.gsub(/^HTTP_/, '').tap(&:downcase!).tap { |s| s.tr!('_', '-') }] = v if k =~ /^HTTP_/
```
The test indicates that you are keeping all of the values, therefore in my opinion:

1. There should also be a test for mixed-case string input, e.g. HTTP_Foo
2. I think the logic as currently proposed will result in the mixed case being preserved which I think is not the right behavior

vpellan replied:
As you pointed in your second comment, HTTP header names are uppercased by Rack (as HTTP header names are case-insensitive according to RFC 2616) and Rack will also drop empty headers so 'HTTP_' should not be possible either. But there's the case of someone adding a middleware to Rack that changes the headers for exemple. I believe that in these case, we still want to send everything to the WAF, as an attack could still be present in the value of these headers.

p-datadog commented:
Given the original report was in reference to invalid headers being sent to the application by external clients, would it also make sense to check for *duplicate* headers being sent? Including in varying case? E.g. HTTP_FOO and HTTP_Foo in the same request.

vpellan commented:
> Given the original report was in reference to invalid headers being sent to the application by external clients, would it also make sense to check for _duplicate_ headers being sent? Including in varying case? E.g. HTTP_FOO and HTTP_Foo in the same request.

I added support for duplicate headers, including in varying case, as stated by [RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2), header keys are case insensitive, so HTTP_FOO and HTTP_Foo are the same header key, and thus it is semantically equal to HTTP_FOO: "value1, value2" (for HTTP_FOO: value1 and HTTP_Foo: value2). It is also possible to send the values as an array to the WAF but that would mean replacing all the strings by arrays in tests.

> And I'm thinking HTTP_FOO_2 and HTTP_2_FOO are also interesting cases to check/add to the test suite, and these should be actually obtainable in real world.

Thank you for that suggestion ! I added these tests too

p-datadog commented on lib/datadog/appsec/contrib/rack/gateway/request.rb:
```diff
@@ -41,7 +41,13 @@ def method
+                # When multiple headers with the same name are present, they are concatenated with a comma
+                # https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
+                # Because headers are case insensitive, HTTP_FOO and HTTP_Foo is the same, and should be merged
+                if k =~ /^HTTP_/
```
Can use `start_with?`

p-datadog commented on lib/datadog/appsec/contrib/rack/gateway/request.rb:
```diff
@@ -41,7 +41,13 @@ def method
+                # Because headers are case insensitive, HTTP_FOO and HTTP_Foo is the same, and should be merged
+                if k =~ /^HTTP_/
+                  key = k.gsub(/^HTTP_/, '').tap(&:downcase!).tap { |s| s.tr!('_', '-') }
+                  h[key] = h[key].nil? ? v : "#{h[key]}, #{v}"
```
Assigning `h[key]` to a variable will save one hash access

p-datadog reviewed (approved):
I see several optimization opportunities, don't know how performance-critical you think this code is. Functionally I think it's good to go.

lloeki commented on lib/datadog/appsec/contrib/rack/gateway/request.rb:
```diff
@@ -41,7 +41,14 @@ def method
-                h[k.gsub(/^HTTP_/, '').downcase!.tr('_', '-')] = v if k =~ /^HTTP_/
+                # When multiple headers with the same name are present, they are concatenated with a comma
+                # https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
+                # Because headers are case insensitive, HTTP_FOO and HTTP_Foo is the same, and should be merged
```
When could this happen?

- The Rack means upcasing HTTP headers into the `HTTP_UPCASED_HEADER` variant
- Given that `request.env.each_with_object` iterates on the rack env (which is a hash) there can be no duplicates except for mixed case, which cannot happen when Rack compliant due to the above.
- The Rack env does not concatenate multiple identical headers into a single hash key, instead making use of a single one (can't recall if first or last).

It follows that the only thing that could fill in some `HTTP_MiXeD_CaSe` is some non-Rack compliant middleware or monkeypatch that  inject pseudo headers that actually don't exist on the HTTP request AND would execute before Datadog's middlewares, which are designed to be run first at the top of the Rack stack.

Note: According to the HTTP spec concatenation MAY happen but MUST only happen for headers that are lists. As is this code does so for ALL headers so it's very much incorrect, and possibly prone to injections or opportunities for a WAF bypass (e.g `Authorization` headers being concatenated in a way that make it valid for libddwaf rules but would in isolation - as seen by the application - be dangerous)

vpellan replied:
I agree that this is an extremely rare edge case, that would be introduced by some sort of weird middleware, and I believe that we should not be responsible for a client modifying the core behaviour of Rack, especially when it is only hypothetical. And a quick benchmark shows that this is 50x slower than your first suggestion with a bit of modification (replacing gsub by delete_prefix and regex in condition by start_with), which I believe is more important for our clients.
